harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -116,6 +171,127 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
+
+            if self.state.last_ckpt != self.steps_completed:
+                if self.ckpt_policy == "all" or (
+                    self.ckpt_policy == "best"
+                    and self.is_best_validation(now=searcher_metric,
+                                                before=self.core_context.train.get_experiment_best_validation())
+                ):
+                    self.checkpoint(already_exiting=False)
+
+    def is_best_validation(self, now: float, before: Optional[float]) -> bool:
+        if before is None:
+            return True
+        smaller_is_better = self.env.experiment_config["searcher"]["smaller_is_better"]
+        return (now < before) if smaller_is_better else (now > before)
+
+    def on_epoch_start(self, epoch_idx: int):
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_start"
+            ):
+                sig = signature(callback.on_training_epoch_start)
+                if sig.parameters:
+                    callback.on_training_epoch_start(epoch_idx)
+                else:
+                    logging.warning(
+                        "on_training_epoch_start() without parameters is deprecated"
+                        " since 0.17.8. Please add epoch_idx parameter."
+                    )
+                    callback.on_training_epoch_start()  # type: ignore[call-arg]
+
+    def on_epoch_end(self, epoch_idx: int):
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_end"
+            ):
+                callback.on_training_epoch_end(epoch_idx)
+
+    def checkpoint(self, already_exiting: bool):
+        self.core_context.train.set_status("checkpointing")
+        self.last_checkpoint = self.steps_completed
+
+        uuid = ""
+        try:
+            if self.is_chief:
+                metadata = {
+                    "determined_version": det.__version__,
+                    "steps_completed": self.steps_completed,
+                    "framework": f"torch-{torch.__version__}",
+                    "format": "pickle",
+                }
+                with self.context._core.checkpoint.store_path(metadata) as (
+                    path,
+                    storage_id,
+                ):
+                    self._save(path)
+                    uuid = storage_id
+            uuid = self.context.distributed.broadcast(uuid)
+            for callback in self.callbacks.values():
+                callback.on_checkpoint_upload_end(uuid=uuid)
+        except det.InvalidHP:
+            self.core_context.train.report_early_exit(core.EarlyExitReason.INVALID_HP)
+            if not already_exiting:
+                raise ShouldExit(skip_exit_checkpoint=True)
+
+    def _prepare_metrics(self, batch_metrics: List):
```

Comment:
This is a confusing helper method because:
1. It's only called in one place (`report_training_metrics`)
2. It's named as if it were useful in other places
3. It has `for_training=True`, so it's _clearly_ not useful except for in `report_training_metrics`
4. It's physically very far from `report_training_metrics`

Reply:
removed it